File: apps/partner-ios/JYTPartner/DesignModels.swift
import Foundation

// Design + production-run models, mirrored from the partner API:
//   GET /partners/designs            (apps/backend/src/api/partners/designs/route.ts)
//   GET /partners/designs/:id        ([designId]/route.ts — full design +
//                                    partner_info derived from production runs)
//   GET /partners/production-runs    (list, design_id filter)
//   GET /partners/production-runs/:id (detail + tasks)
// and the design module's own columns (modules/designs/models/design.ts):
// media_files json, moodboard scene json, colors + size_sets relations.

// MARK: - Designs

struct PartnerDesignRow: Codable, Identifiable, Hashable {
  let id: String
  var name: String?
  var status: String?
  var thumbnail_url: String?
  var created_at: Date?
  var updated_at: Date?
  var partner_info: PartnerInfo?
  /// owned | assigned | shared — derived server-side, never re-derived here.
  var partner_engagement: String?
  var has_partner_run: Bool?
  var partner_run_count: Int?
}

struct PartnerDesignListResponse: Codable {
  let designs: [PartnerDesignRow]
  let count: Int
  let limit: Int
  let offset: Int
}

/// The partner_status block both design routes emit. Derived entirely from
/// the partner's production runs on the design — the completion timestamps
/// included (apps/backend/src/api/partners/designs/[designId]/route.ts:172).
struct PartnerInfo: Codable, Hashable {
  var partner_status: String?
  var partner_phase: String?
  var partner_started_at: Date?
  var partner_finished_at: Date?
  var partner_completed_at: Date?
  var workflow_tasks_count: Int?
}

/// media_files json on the design: `[{id?, url, isThumbnail?}]`.
struct MediaFile: Codable, Identifiable, Hashable {
  var id: String?
  var url: String
  var isThumbnail: Bool?
}

/// An Excalidraw moodboard scene, kept deliberately loose — the mobile app
/// reads only the embedded images (files[key].dataURL referenced by image
/// elements), never the geometry.
struct MoodboardScene: Codable, Hashable {
  struct Element: Codable, Hashable {
    var type: String?
    var fileId: String?
  }
  struct File: Codable, Hashable {
    var dataURL: String?
    var mimeType: String?
  }
  var elements: [Element]?
  var files: [String: File]?

  /// (key, dataURL) pairs for every image element on the board.
  var imageDataURLs: [(id: String, data: String)] {
    guard let elements = elements, let files = files else { return [] }
    return elements.compactMap { element in
      guard element.type == "image", let key = element.fileId,
            let data = files[key]?.dataURL else { return nil }
      return (key, data)
    }
  }
}

struct DesignColorRow: Codable, Identifiable, Hashable {
  let id: String
  var name: String
  var hex_code: String
  var usage_notes: String?
  var order: Int?
}

struct DesignSizeSetRow: Codable, Identifiable, Hashable {
  let id: String
  var size_label: String
  /// measurement point -> value, stored as JSON; values may be numbers.
  var measurements: [String: FlexibleDouble]?
}

/// An inventory item linked to the design (its bill of materials) — the
/// material options the Complete form offers. The design detail read
/// returns these with raw_materials + location_levels expanded; we only
/// need the naming fields.
struct DesignInventoryItem: Codable, Identifiable, Hashable {
  let id: String
  var title: String?
  var sku: String?
  var unit_of_measure: String?

  var displayName: String { title ?? sku ?? "Material" }
}

/// The full design record from GET /partners/designs/:id. Fields the
/// partner-UI treats as Record<string, any> are decoded leniently here —
/// everything optional, unknown keys dropped.
struct DesignDetail: Codable, Identifiable, Hashable {
  let id: String
  var name: String?
  var description: String?
  var status: String?
  var design_type: String?
  var product_type: String?
  var concept_theme: String?
  var aesthetic_keywords: [String]?
  var designer_notes: String?
  var thumbnail_url: String?
  var media_files: [MediaFile]?
  var moodboard: MoodboardScene?
  var colors: [DesignColorRow]?
  var size_sets: [DesignSizeSetRow]?
  var inventory_items: [DesignInventoryItem]?
  var estimated_cost: FlexibleDouble?
  var cost_currency: String?
  var partner_info: PartnerInfo?
  var is_owner: Bool?
  var created_at: Date?
  var updated_at: Date?
}

/// One uploaded file, as `POST /partners/production-runs/:id/media` and the
/// design media route return them.
struct UploadedFile: Codable, Hashable {
  var id: String?
  var url: String
}

struct UploadFilesResponse: Codable {
  let files: [UploadedFile]
}

/// The attach payload's media entry (`.../media/attach`).
struct AttachMediaFile: Codable {
  var id: String?
  var url: String
  var isThumbnail: Bool?
}

struct AttachMediaBody: Codable {
  var media_files: [AttachMediaFile]
}

// MARK: - Production runs

struct ProductionRun: Codable, Identifiable, Hashable {
  let id: String
  var status: String?
  var run_type: String?
  var role: String?
  var quantity: Int?
  var design_id: String?
  var accepted_at: Date?
  var started_at: Date?
  var finished_at: Date?
  var completed_at: Date?
  var created_at: Date?
  var updated_at: Date?
}

struct ProductionRunListResponse: Codable {
  let production_runs: [ProductionRun]
  let count: Int
  let limit: Int
  let offset: Int
}

struct RunTask: Codable, Identifiable, Hashable {
  let id: String?
  var title: String?
  var status: String?
  var priority: String?
  var start_date: Date?
  var end_date: Date?
}

struct ProductionRunDetail: Codable {
  let production_run: ProductionRun
  var tasks: [RunTask]?
}

// MARK: - Lenient JSON scalars

/// bigNumber columns and other numbers that may arrive as Double, Int, or
/// numeric String — Medusa's query.graph is inconsistent here.
struct FlexibleDouble: Codable, Hashable {
  let value: Double?

  init(from decoder: Decoder) throws {
    let container = try decoder.singleValueContainer()
    if let d = try? container.decode(Double.self) {
      value = d
    } else if let i = try? container.decode(Int.self) {
      value = Double(i)
    } else if let s = try? container.decode(String.self) {
      value = Double(s)
    } else {
      value = nil
    }
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.singleValueContainer()
    try container.encode(value)
  }
}

File: apps/partner-android/app/src/main/java/com/jyt/partner/models/DesignModels.kt
package com.jyt.partner.models

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonPrimitive

// Design + production-run models — the Kotlin mirror of the iOS app's
// DesignModels.swift and the backend /partners/designs + /partners/production-runs
// surfaces.

@Serializable
data class PartnerDesignRow(
    val id: String,
    val name: String? = null,
    val status: String? = null,
    @SerialName("thumbnail_url") val thumbnailUrl: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
    @SerialName("partner_info") val partnerInfo: PartnerInfo? = null,
    /** owned | assigned | shared — derived server-side, never re-derived here. */
    @SerialName("partner_engagement") val partnerEngagement: String? = null,
    @SerialName("has_partner_run") val hasPartnerRun: Boolean? = null,
    @SerialName("partner_run_count") val partnerRunCount: Int? = null,
)

@Serializable
data class PartnerDesignListResponse(
    val designs: List<PartnerDesignRow>,
    val count: Int,
    val limit: Int,
    val offset: Int,
)

/// The partner_status block both design routes emit — derived entirely from
/// the partner's production runs on the design.
@Serializable
data class PartnerInfo(
    @SerialName("partner_status") val partnerStatus: String? = null,
    @SerialName("partner_phase") val partnerPhase: String? = null,
    @SerialName("partner_started_at") val partnerStartedAt: String? = null,
    @SerialName("partner_finished_at") val partnerFinishedAt: String? = null,
    @SerialName("partner_completed_at") val partnerCompletedAt: String? = null,
    @SerialName("workflow_tasks_count") val workflowTasksCount: Int? = null,
)

/** media_files json on the design: `[{id?, url, isThumbnail?}]`. */
@Serializable
data class MediaFile(
    val id: String? = null,
    val url: String,
    @SerialName("isThumbnail") val isThumbnail: Boolean? = null,
)

/** An Excalidraw moodboard scene, kept loose — only the embedded images. */
@Serializable
data class MoodboardScene(
    val elements: List<Element>? = null,
    val files: Map<String, File>? = null,
) {
    @Serializable
    data class Element(val type: String? = null, @SerialName("fileId") val fileId: String? = null)

    @Serializable
    data class File(@SerialName("dataURL") val dataUrl: String? = null)

    /** (key, dataURL) pairs for every image element on the board. */
    val imageDataUrls: List<Pair<String, String>>
        get() = elements.orEmpty().mapNotNull { element ->
            val key = element.fileId ?: return@mapNotNull null
            val data = files?.get(key)?.dataUrl ?: return@mapNotNull null
            if (element.type == "image") key to data else null
        }
}

@Serializable
data class DesignColorRow(
    val id: String,
    val name: String,
    @SerialName("hex_code") val hexCode: String,
    @SerialName("usage_notes") val usageNotes: String? = null,
    val order: Int? = null,
)

@Serializable
data class DesignSizeSetRow(
    val id: String,
    @SerialName("size_label") val sizeLabel: String,
    /// measurement point -> value; values may be numbers or strings.
    val measurements: Map<String, FlexibleDouble>? = null,
)

/** An inventory item linked to the design (its bill of materials). */
@Serializable
data class DesignInventoryItem(
    val id: String,
    val title: String? = null,
    val sku: String? = null,
    @SerialName("unit_of_measure") val unitOfMeasure: String? = null,
) {
    val displayName: String get() = title ?: sku ?: "Material"
}

@Serializable
data class DesignDetail(
    val id: String,
    val name: String? = null,
    val description: String? = null,
    val status: String? = null,
    @SerialName("design_type") val designType: String? = null,
    @SerialName("product_type") val productType: String? = null,
    @SerialName("concept_theme") val conceptTheme: String? = null,
    @SerialName("aesthetic_keywords") val aestheticKeywords: List<String>? = null,
    @SerialName("designer_notes") val designerNotes: String? = null,
    @SerialName("thumbnail_url") val thumbnailUrl: String? = null,
    @SerialName("media_files") val mediaFiles: List<MediaFile>? = null,
    val moodboard: MoodboardScene? = null,
    val colors: List<DesignColorRow>? = null,
    @SerialName("size_sets") val sizeSets: List<DesignSizeSetRow>? = null,
    @SerialName("inventory_items") val inventoryItems: List<DesignInventoryItem>? = null,
    @SerialName("estimated_cost") val estimatedCost: FlexibleDouble? = null,
    @SerialName("cost_currency") val costCurrency: String? = null,
    @SerialName("partner_info") val partnerInfo: PartnerInfo? = null,
    @SerialName("is_owner") val isOwner: Boolean? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
)

@Serializable
data class DesignDetailResponse(val design: DesignDetail)

@Serializable
data class UploadFile(
    val id: String? = null,
    val url: String,
)

@Serializable
data class UploadFilesResponse(val files: List<UploadFile>)

@Serializable
data class AttachMediaFile(
    val id: String? = null,
    val url: String,
    @SerialName("isThumbnail") val isThumbnail: Boolean? = null,
)

@Serializable
data class AttachMediaBody(
    @SerialName("media_files") val mediaFiles: List<AttachMediaFile>,
)

// MARK: Production runs

@Serializable
data class ProductionRun(
    val id: String,
    val status: String? = null,
    @SerialName("run_type") val runType: String? = null,
    val role: String? = null,
    val quantity: Int? = null,
    @SerialName("design_id") val designId: String? = null,
    @SerialName("accepted_at") val acceptedAt: String? = null,
    @SerialName("started_at") val startedAt: String? = null,
    @SerialName("finished_at") val finishedAt: String? = null,
    @SerialName("completed_at") val completedAt: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
)

@Serializable
data class ProductionRunListResponse(
    @SerialName("production_runs") val productionRuns: List<ProductionRun>,
    val count: Int,
    val limit: Int,
    val offset: Int,
)

@Serializable
data class RunTask(
    val id: String? = null,
    val title: String? = null,
    val status: String? = null,
    val priority: String? = null,
    @SerialName("start_date") val startDate: String? = null,
    @SerialName("end_date") val endDate: String? = null,
)

@Serializable
data class ProductionRunDetail(
    @SerialName("production_run") val productionRun: ProductionRun,
    val tasks: List<RunTask>? = null,
)

// MARK: Lenient JSON scalars

/**
 * bigNumber columns and other numbers that may arrive as Double, Int, or
 * numeric String — Medusa's query.graph is inconsistent here.
 */
@Serializable(with = FlexibleDoubleSerializer::class)
class FlexibleDouble(val value: Double?)

object FlexibleDoubleSerializer : KSerializer<FlexibleDouble> {
    private val surrogate = JsonElement.serializer()
    override val descriptor: SerialDescriptor = surrogate.descriptor

    override fun serialize(encoder: Encoder, value: FlexibleDouble) {
        val v = value.value
        if (v == null) encoder.encodeNull()
        else encoder.encodeSerializableValue(surrogate, JsonPrimitive(v))
    }

    override fun deserialize(decoder: Decoder): FlexibleDouble {
        val element = decoder.decodeSerializableValue(surrogate)
        return FlexibleDouble(
            runCatching { element.jsonPrimitive.doubleOrNull }.getOrNull()
        )
    }
}
